### DISCUSSION TITLE
Table and Form Resources as a Partial

### DISCUSSION THREAD

**Discussion by caiotarifa** [activeadmin/activeadmin#8236] [Ideas]
Hello,

I'm the creator of [Formadmin](https://github.com/formaweb/formadmin), a theme for the old version of ActiveAdmin.

I've been delving into ActiveAdmin v4 and I'm really impressed with the updates, particularly the integration of TailwindCSS and the enhanced view customization options. This is already a significant step forward, and I want to congratulate and thank you for this fantastic effort.

**Are there any plans to introduce partials (`.erb`) for table and form resources, allowing for further customization?**

---

I understand that it's possible to modify their style via the `plugin.js` file, but I'm looking for more—specifically, the ability to rearrange and customize classes directly in the `.erb` file.

**Comment by javierjulio** [comment]
Wow, thank you so much. I really appreciate that. I’m glad you like the update. Can you be specific about what view elements in partials you’re looking for? The form is already available just not considered public. The table may be difficult because it’s a component. I’m open to improvements here. TailwindCSS classes written directly into templates, those templates are considered public and should be already included in the views command to copy to your project. The only styles in the plugin are for components, stuff you wouldn’t have direct access to but can still style. They are written with low specificity so it should be easier to override and would work with Tailwind utilities. For example, using min-w-* on the index table column. Note that you can just write normal CSS in your stylesheet to override the rules in the plugin. No need to modify the plugin itself. You could, but should be simple enough to override with your own CSS. That’s very much by design.

**Reply by javierjulio** [reply]
The sidebar and active filters are already in a partial. I have considered a partial for filters form but it’s just a few lines. You can already modify the placement of the sidebar yourself. The filters form you could too if we had that be in a partial which we could do.

**Reply by caiotarifa** [reply]
Thank you, @javierjulio. I appreciate your response and the time you've taken.

As a front-end developer, my preference is to have complete access and control over all the rendered HTML, not just the styles.

I believe the `plugin.js` file should be simplified to only include global theme settings. For everything else, my suggestion is to create a components folder within the views, something like this:

```
views/
├─ active_admin/
│  ├─ components/
│  │  ├─ _button.html.erb
│  │  ├─ _scopes.html.erb
│  │  ├─ _table.html_erb
│  │  ├─ _form.html.erb
│  │  ├─ _filter.html.erb
```

HTML and CSS (classes) could be contained in these files instead of using `@apply`. This approach opens up a world of customization possibilities, encourages code reuse, and aligns more closely with the architecture of frameworks like React and Vue.

For example, in some screens, I need to list numerous filters, so I'd like to add an 'accordion' feature to group them. This would be easier if I could access and modify the HTML directly (in the old version, I had to create the necessary elements using JavaScript, which was quite a workaround).

![accordion-filter](https://github.com/activeadmin/activeadmin/assets/197979/96d44ac1-8499-4f29-809c-eaf1c7cd6192)

---

_You said the filters are already in partials, but no, there is a partial for the sidebar that imports all sections. In my case, I would like to isolate and modify only the filter section, keeping the other sections in the sidebar._

**Reply by javierjulio** [reply]
No I said the active filters are in a partial. I can see though how the name is confusing. It’s what we call the sidebar section that shows the dynamic text of what filters are applied. The filters form, which you are asking about, is what I was referring to not being in a partial. We can consider that though. I’m not sure if the way we render filters would give you that kind of access. But I guess since it’s in Ruby you could group them yourself. In regards to other components, like scopes or index list, I have considered moving those to partials. Not sure how realistic it will be to get all output into partials. My hope is that what is there will be good for most for customizing the admin layout without needing to do monkey patches.

**Reply by caiotarifa** [reply]
Sorry for the mix-up regarding the names, and thanks for clearing that up. It would be great if the filter form could be made into a partial.

I'm also curious about your thoughts on the challenges or limitations of converting and exposing everything as ERB partials. Could you share more about why this might not be very realistic?

**Reply by javierjulio** [reply]
No worries at all. Just wanted to clarify. Thanks. I'm not sure based on what you've shared earlier if it's what you would want. Adding a partial for the filters form is simple enough, [the contents are just two lines](https://github.com/activeadmin/activeadmin/blob/master/lib/active_admin/filters/resource_extension.rb#L167-L173). This is [the helper method for building the filters form](https://github.com/activeadmin/activeadmin/blob/master/app/helpers/active_admin/form_helper.rb#L62-L87) which could be it. The problem though is that the actual HTML output is defined in Formtastic which is what I think you'd want to have access to. If that's the case it would have to require some research into say removing Formtastic from filters form and rely on our own form builder and inputs structure which may be worthwhile. You are welcome to investigate that if it's something that interests you. My hunch is it's unlikely to change. It may be a substantial project.

I take it what you mean is more about exposing everything in partials, regardless of whether they are ERB or not. Some partials have to be done in Arbre, with an `.arb` extension. Content in Arbre has to render in its own context and is different from ERB. I've definitely pushed some limits here on moving things to ERB and that will have an effect on upgrading which is noted in the Upgrading guide. I'd recommend trying out the local app we use for development as well and seeing this for yourself. You'll need to have a deeper understanding of the library, specifically Arbre, and it will be clearer why it's not so straightforward to say, put everything in ERB.

Other than customization, one of the primary reasons I migrated so many components to partials is first many have no need to be a component (e.g. rendering layout, site header, action items etc.) and second as a template it helps speed up development as there is no need to restart a dev server since the views are in app directory vs being Ruby in lib.
